compiler/codegen/ELFGenerator.cpp

```diff
@@ -0,0 +1,562 @@
+    shdr->sh_size = shSize;
+    shdr->sh_link = 0;
+    shdr->sh_info = 0;
+    shdr->sh_addralign = 32;
+    shdr->sh_entsize = 0;
+
+    _textSection = shdr;
+    strcpy(_textSectionName, ".text");
+}
+    
+void
+TR::ELFGenerator::initializeDynSymSection(uint32_t shName, ELFOffset shOffset, uint32_t shSize, uint32_t shLink){
+    
+    ELFSectionHeader * shdr = static_cast<ELFSectionHeader *>(_rawAllocator.allocate(sizeof(ELFSectionHeader)));
+
+    shdr->sh_name = shName;
+    shdr->sh_type = SHT_SYMTAB; // SHT_DYNSYM
+    shdr->sh_flags = 0; //SHF_ALLOC;
+    shdr->sh_addr = 0; //(ELFAddress) &((uint8_t *)_elfHeader + symbolStartOffset); // fake address because not continuous
+    shdr->sh_offset = shOffset;
+    shdr->sh_size = shSize;
+    shdr->sh_link = shLink; // dynamic string table index
+    shdr->sh_info = 1; // index of first non-local symbol: for now all symbols are global
+    shdr->sh_addralign = 8;
+    shdr->sh_entsize = sizeof(ELFSymbol);
+
+    _dynSymSection = shdr;
+    strcpy(_dynSymSectionName, ".symtab");
+}
+
+void
+TR::ELFGenerator::initializeStrTabSection(uint32_t shName, ELFOffset shOffset, uint32_t shSize){
+    
+    ELFSectionHeader * shdr = static_cast<ELFSectionHeader *>(_rawAllocator.allocate(sizeof(ELFSectionHeader)));
+    
+    shdr->sh_name = shName;
+    shdr->sh_type = SHT_STRTAB;
+    shdr->sh_flags = 0;
+    shdr->sh_addr = 0;
+    shdr->sh_offset = shOffset;
+    shdr->sh_size = shSize;
+    shdr->sh_link = 0;
+    shdr->sh_info = 0;
+    shdr->sh_addralign = 1;
+    shdr->sh_entsize = 0;
+
+    _shStrTabSection = shdr;
+    strcpy(_shStrTabSectionName, ".shstrtab");
+}
+
+void
+TR::ELFGenerator::initializeDynStrSection(uint32_t shName, ELFOffset shOffset, uint32_t shSize){
+
+    ELFSectionHeader * shdr = static_cast<ELFSectionHeader *>(_rawAllocator.allocate(sizeof(ELFSectionHeader)));
+
+    shdr->sh_name = shName;
+    shdr->sh_type = SHT_STRTAB;
+    shdr->sh_flags = 0;
+    shdr->sh_addr = 0;
+    shdr->sh_offset = shOffset;
+    shdr->sh_size = shSize;
+    shdr->sh_link = 0;
+    shdr->sh_info = 0;
+    shdr->sh_addralign = 1;
+    shdr->sh_entsize = 0;
+
+    _dynStrSection = shdr;
+    strcpy(_dynStrSectionName, ".dynstr");
+}
+
+void
+TR::ELFGenerator::initializeRelaSection(uint32_t shName, ELFOffset shOffset, uint32_t shSize){
+
+    ELFSectionHeader * shdr = static_cast<ELFSectionHeader *>(_rawAllocator.allocate(sizeof(ELFSectionHeader)));
+
+    shdr->sh_name = shName;
+    shdr->sh_type = SHT_RELA;
+    shdr->sh_flags = 0;
+    shdr->sh_addr = 0;
+    shdr->sh_offset = shOffset;
+    shdr->sh_size = shSize;
+    shdr->sh_link = 3; // dynsymSection index in the elf file. kept this hardcoded for now as this shdr applies to relocatable only
+    shdr->sh_info = 1;
+    shdr->sh_addralign = 8;
+    shdr->sh_entsize = sizeof(ELFRela);
+
+    _relaSection = shdr;
+    strcpy(_relaSectionName, ".rela.text");
+
+}
+
+void
+TR::ELFGenerator::buildELFFile(::FILE *elfFile){
```

Comment:
I thought of using the name `buildELFFile` mainly because it was putting all the pieces (building blocks) of the ELF file together. I think `emitELFFile` would be fine too.